dotCMS/src/main/java/com/dotcms/rest/api/v1/vtl/VTLResource.java

```diff
@@ -60,86 +64,127 @@ public VTLResource() {
+                    + "' under path '" + VTL_PATH + StringPool.SLASH + folderName + StringPool.SLASH + "'";
+            Logger.error(this, errorMessage, e);
+            return ResponseUtil.mapExceptionResponse(new DotDataException(errorMessage));
+        } catch(Exception e) {
+            Logger.error(this,"Exception on VTL endpoint. POST method: " + e.getMessage(), e);
+            return ResponseUtil.mapExceptionResponse(e);
         }
+    }
+
+    private Response evalVTLFile(final HttpServletRequest request, final HttpServletResponse response,
+                                 final FileAsset getFileAsset, final Map<String, Object> contextParams,
+                                 final User user, final DotJSONCacheStrategy cacheStrategy)
+            throws IOException {
+        final org.apache.velocity.context.Context context = VelocityUtil.getInstance().getContext(request, response);
+        contextParams.forEach(context::put);
+        context.put("dotJSON", new DotJSON());
+
+        final StringWriter evalResult = new StringWriter();
+
+        try (final InputStream fileAssetIputStream = getFileAsset.getInputStream()) {
+            VelocityUtil.getEngine().evaluate(context, evalResult, "", fileAssetIputStream);
+            DotJSON<String, String> dotJSON = (DotJSON<String, String>) context.get("dotJSON");
 
-        final String requestURI = request.getRequestURI() + StringPool.QUESTION + request.getQueryString();
+            if(dotJSON.size()==0) { // If dotJSON is not used let's return the raw evaluation of the velocity file
+                return Response.ok(evalResult.toString()).build();
+            } else {
+                // let's add it to cache
+                cacheStrategy.addIfNeeded(request, user, dotJSON);
+                return Response.ok(dotJSON).build();
+            }
+        }
+    }
 
-        return new DotJSONCacheKey(initDataObject.getUser(), language, requestURI, persona);
+    private FileAsset getVTLFile(final HTTPMethod httpMethod, HttpServletRequest request, String folderName,
+                                 final User user) throws DotDataException, DotSecurityException {
+        Language currentLanguage = WebAPILocator.getLanguageWebAPI().getLanguage(request);
+        final Host site = this.hostAPI.resolveHostName(request.getServerName(), APILocator.systemUser(), false);
+        final String vtlFilePath = VTL_PATH + StringPool.SLASH + folderName + StringPool.SLASH
+                + httpMethod.fileName + FILE_EXTENSION;
+        final Identifier identifier = identifierAPI.find(site, vtlFilePath);
+        final Contentlet getFileContent = contentletAPI.findContentletByIdentifier(identifier.getId(), true,
+                currentLanguage.getId(), user, true);
+        return APILocator.getFileAssetAPI().fromContentlet(getFileContent);
     }
 
-    private enum HTTPMethod {
+    public enum HTTPMethod {
```

Comment:
this looks something we can reuse outside to this class, it would be good to have it out of here